### ISSUE TITLE
CSV loading start using swap

### ISSUE THREAD

**Issue by chriskamphuis** [duckdb/duckdb#321]
I am trying to load a big CSV file into the database. ~100.000.000 lines of three integers. The file has a size of 1.5G. When I try to load this file, duckdb starts using swap and basically my machine (8G of ram) can not handle it. MonetDB can load this data without issues.

If I split the file in multiple files of 1.000.000 lines, and load these files separately, everything goes fine. However, duckDB is much slower when querying over this data than MonetDB. I suspect because I load the data in multiple iterations DuckDB might assume the data is not sorted anymore (while it is)?

I think duckDB should be able to load the data no problem?, and am I right in assuming DuckDB becomes slower due the fact that I do load my data in iterations?

**Comment by hannesmuehleisen** [comment]
This should be solved once the `buffermanager` branch is merged, so stand by please.

**Comment by hannesmuehleisen** [comment]
> If I split the file in multiple files of 1.000.000 lines, and load these files separately, everything goes fine. However, duckDB is much slower when querying over this data than MonetDB. I suspect because I load the data in multiple iterations DuckDB might assume the data is not sorted anymore (while it is)?

No, that's not the case. Its likely that the speed difference is due to the fact that we have not yet optimised for speed at all.

> I think duckDB should be able to load the data no problem?, and am I right in assuming DuckDB becomes slower due the fact that I do load my data in iterations?

No, that should make no difference either.

**Comment by Mytherin** [comment]
Thanks for the feedback!

> I am trying to load a big CSV file into the database. ~100.000.000 lines of three integers. The file has a size of 1.5G. When I try to load this file, duckdb starts using swap and basically my machine (8G of ram) can not handle it. MonetDB can load this data without issues.

Could you perhaps share a representative (possibly anonymized) sample of the data for us to check? As Hannes mentioned the swapping should indeed be fixed by the buffer manager branch, but it is always interesting for us to see which data sets people are using so we can use them for testing.


> If I split the file in multiple files of 1.000.000 lines, and load these files separately, everything goes fine. However, duckDB is much slower when querying over this data than MonetDB. I suspect because I load the data in multiple iterations DuckDB might assume the data is not sorted anymore (while it is)?

No, that should not make a difference. DuckDB does not perform any optimizations if the data is inherently sorted. MonetDB does perform these. In case you want to perform point queries efficiently (e.g. `SELECT x FROM tbl WHERE y=10`) perform a `CREATE INDEX` statement first. Otherwise it might help us if you could share a sample of your workload as well so we can see what the problem could be.

**Comment by Chriskamphuis** [comment]
Thanks for the quick answers.

>  Could you perhaps share a representative (possibly anonymized) sample of the data for us to check? As Hannes mentioned the swapping should indeed be fixed by the buffer manager branch, but it is always interesting for us to see which data sets people are using so we can use them for testing.

Yes, the data can be downloaded here: [terms](https://www.cs.ru.nl/~arjen/tmp/terms-robust.csv.xz), [docs](https://www.cs.ru.nl/~arjen/tmp/docs-robust.csv.xz), and [dict](https://www.cs.ru.nl/~arjen/tmp/dict-robust.csv.xz). We are doing queries similar as described in [this](https://hannes.muehleisen.org/SIGIR2014-column-stores-ir-prototyping.pdf) paper. 

> No, that should not make a difference. DuckDB does not perform any optimizations if the data is inherently sorted. MonetDB does perform these. In case you want to perform point queries efficiently (e.g. `SELECT x FROM tbl WHERE y=10`) perform a CREATE INDEX statement first. Otherwise it might help us if you could share a sample of your workload as well so we can see what the problem could be.

Ok, I assumed it did. Good to know that is not the case. I will first try to create indexes on the columns, and see what the effect of this will be.

**Comment by Mytherin** [comment]
Thanks for the links, unfortunately I'm getting a 404 when I follow them. Could you check if they work for you?

**Comment by Chriskamphuis** [comment]
They are down, I believe @arjenpdevries took them down because he prefered to not keep them online and @hannesmuehleisen already downloaded them.

**Comment by Mytherin** [comment]
Ah great, no worries then. I will get them from Hannes. Thanks again.

**Comment by hannesmuehleisen** [comment]
@Chriskamphuis  The buffermanager branch has been merged, can you try to see whether this is fixed?

**Comment by Mytherin** [comment]
Note that by default the memory limit is currently infinite, you can set a different limit using eg PRAGMA memory_limit=8GB

**Comment by Chriskamphuis** [comment]
I tried it for a 1.5G file, this now works without any issues.

I also tried it for a 6.8G file, this one still fails. However it kills the python process instead of freezing the computer (which happened before).

I am not entirely sure how the PRAGMA statement should work? If I for example try these commands:
```
>>> c.execute('PRAGMA memory_limit=4GB')
```

```
>>> c.execute('PRAGMA enable_profiling')
```

I get the following error:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
duckdb.DatabaseError: Cannot prepare multiple statements at once!
```

**Comment by Chriskamphuis** [comment]
I suppose that #342 fixes this? When its merged I will retry it.

**Comment by hannesmuehleisen** [comment]
The `PRAGMA` currently does not work from the Python client unfortunately.. You can try with the shell?

**Comment by Chriskamphuis** [comment]
After issuing `PRAGMA memory_limit=4GB` it still starts using swap after a couple of minutes.

**Comment by hannesmuehleisen** [comment]
I can see that, it still grows, just slower. Indeed something to investigate.

**Comment by hannesmuehleisen** [comment]
@Chriskamphuis if you load in multiple parts with `COMMIT` in between, do you still see the issue?

**Comment by Chriskamphuis** [comment]
If I load the file in multiple parts it already works in general.

**Comment by Chriskamphuis** [comment]
Just to confirm, loading everything in multiple parts with explicitly commiting the parts shows the same behaviour. (It loads fine). So only the problem when loading everything at once remains.

**Comment by hannes** [comment]
Duplicate of #825

**Comment by hannes** [comment]
Hey @chriskamphuis can you try again whether the memory limit is still an issue if you use either a persistent database file or explicitly set a temporary directory?

**Comment by kpm** [comment]
Hi, I have a similar problem when loading an 11GB csv file into a new on-disk duckdb database.  I'm using computer with 16G Ram on ubuntu 20.04.  With the python client version 0.3.5-dev699, if I set the memory limit to 12GB, the script still uses 15g of resident memory as reported by top.

**Comment by Alex-Monahan** [comment]
> Hi, I have a similar problem when loading an 11GB csv file into a new on-disk duckdb database. I'm using computer with 16G Ram on ubuntu 20.04. With the python client version 0.3.5-dev699, if I set the memory limit to 12GB, the script still uses 15g of resident memory as reported by top.

Hello! Could you provide your query, and if possible, a way to generate a representative csv file?

**Comment by kpm** [comment]
These are the queries executed from my python script:

```
con.execute("""create table comments(
by VARCHAR not null,
id integer not null primary key,
parent integer not null,
text varchar null,
time TIMESTAMPtz not null,
type varchar not null)""")

print("loading comments into duckdb")
con.execute("COPY comments(by,id,parent,text,time,type) FROM '"+ COMMENTS_PROCESSED_CSV + "' header")
```

I have attached [comments_head.csv](https://github.com/duckdb/duckdb/files/8829125/comments_head.csv) the head of the 11Gb csv file, does that help at all?

**Comment by chriskamphuis** [comment]
For me it also still happens for latest version of DuckDB, also with persistent database, using a 30GB  csv file while having 16GB RAM.

**Comment by chriskamphuis** [comment]
Any update on this maybe? It still happens for me on Ubuntu 22 with 16 GB RAM on a persistent file, if I try to load [this](https://rgw.cs.uwaterloo.ca/JIMMYLIN-bucket0/mmead/enwiki-20190701-wiki2vec-dim300.tar.bz2) data (after extraction) of size 30GB the following way: 

```
import duckdb

cursor = duckdb.connect('db').cursor()
cursor.execute("CREATE TEMP SEQUENCE identifiers;")
cursor.execute(f"""CREATE OR REPLACE TABLE embeddings AS SELECT nextval('identifiers') AS id, column000 AS key, {str(['column{:0>{}}'.format(i, 3) for i in range(1,301)]).replace("'", "")} AS embedding FROM read_csv_auto('/path/to/data', skip=1, delim=' ')""")
cursor.execute("DROP SEQUENCE identifiers")
```
Sorry for the ugly query :-)
I am trying to create a table with a list of 300 doubles.

**Comment by chriskamphuis** [comment]
Unfortunately for me at  0.5.2.dev1142 it also happens

**Comment by Mytherin** [comment]
The file seems to be written to disk optimistically and I don't see memory usage spiking much on my Macbook. Perhaps this is related to the jemalloc issue (#3969). Could you try either (1) setting the memory limit yourself (`SET memory_limit='XGB'`) or (2) using [LD_PRELOAD to force usage of jemalloc](https://github.com/jemalloc/jemalloc/wiki/Getting-Started)?

**Comment by chriskamphuis** [comment]
It works now when setting memory limit, thanks Mark!

**Comment by hannes** [comment]
woo this is closed
